enhancements/installer/assisted-installer-in-cluster.md

```diff
@@ -0,0 +1,580 @@
+---
+title: assisted-installer-in-cluster
+authors:
+  - "@mhrivnak"
+  - "@hardys"
+  - "@avishayt"
+reviewers:
+  - "@eparis"
+  - "@markmc"
+  - "@dgoodwin"
+  - "@ronniel1"
+
+approvers:
+  - TBD
+creation-date: 2020-12-22
+last-updated: 2021-02-10
+status: implementable
+see-also:
+  - enhancements/installer/connected-assisted-installer.md
+  - enhancements/installer/assisted-installer-bare-metal-validations.md
+---
+
+# Assisted Installer in-cluster
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The Assisted Installer currently runs as a SaaS on cloud.redhat.com, enabling
+users to deploy OpenShift clusters with certain customizations, particularly on
+bare metal hardware. It is necessary to bring those capabilities on-premise in
+users' "Hub" clusters by installing clusters via Multi-cluster management, such
+as through Hive and RHACM.
+
+This enhancement proposes the first iteration of running the Assisted Installer
+in end-user clusters to meet the purpose above.
+
+## Motivation
+
+The Assisted Installer running as a SaaS has demonstrated a new and valuable
+way to install OpenShift clusters through an agent-based flow. The [Assisted
+Installer for Connected Environments](connected-assisted-installer.md)
+enhancement proposal elaborates on the value of the agent-based installations.
+
+However many customers do not want a third party service running outside of
+their network to be able to provision hosts inside their network. Reasons
+include trust, control, and reproducibility. Those users will prefer to run the
+service on-premises. There are also many customers whose network environments
+have limited or no connectivity to the internet. Finally, it is desirable to
+integrate this new installation method into RHACM, which runs on-premesis.
+
+### Goals
+
+* Expose the Assisted Installer's capabilities as a kubernetes-native API. Some
+portion may be exposed through [Hive](https://github.com/openshift/hive)'s API.
+* Enable multi-cluster management tooling to create new clusters using Assisted Installer.
+* Enable adding workers to any OpenShift cluster deployed by the Assisted Installer via
```

Comment:
I see three broad topics to cover with an agent-based flow:
1. create a cluster
2. add workers on day 2
3. add remote workers on day 2

I already removed topics specific to point 3 from this proposal and will plan a separate proposal for that.

I think you're suggesting that points 1 and 2 should also be separated. I've been considering that and specifically wrestling with the question of whether point 1 by itself is a sufficiently complete feature to be approved without point 2. Generally speaking, if we enable hive to create clusters but don't provide a way to add workers, is that viable from a user perspective? I could see it as an incomplete feature. Is it even a requirement for an individual proposal to be a complete viable feature? Maybe I'm over-thinking that.

There is also the coincidence that we are primarily interested right now in creation of single-node clusters that will not be scaled with worker nodes. For that use case, it happens that point 1 is viable and useful by itself, though that is based on a specific short-term priority. So I'm not sure that's worth much consideration in an OKD enhancement proposal.

All that said, do you think moving all aspects of adding workers on day 2 into a separate proposal would be the best approach? I'm happy to make that change.

Reply:
> I see three broad topics to cover with an agent-based flow:
> 
> 1. create a cluster
> 2. add workers on day 2
> 3. add remote workers on day 2
> 
> I already removed topics specific to point 3 from this proposal and will plan a separate proposal for that.
> 
> I think you're suggesting that points 1 and 2 should also be separated.

Yes, particularly since (3) in a single-cluster context I think is the next most pressing concern (for telco use cases). But maybe we should follow up separately to talk that through.

> I've been considering that and specifically wrestling with the question of whether point 1 by itself is a sufficiently complete feature to be approved without point 2. Generally speaking, if we enable hive to create clusters but don't provide a way to add workers, is that viable from a user perspective?

Oh wow. Maybe I was being dense, but it wasn't obvious to me we were talking about add workers specifically for clusters created by Hive.

> I could see it as an incomplete feature. Is it even a requirement for an individual proposal to be a complete viable feature? Maybe I'm over-thinking that.
> 
> There is also the coincidence that we are primarily interested right now in creation of single-node clusters that will not be scaled with worker nodes. For that use case, it happens that point 1 is viable and useful by itself, though that is based on a specific short-term priority. So I'm not sure that's worth much consideration in an OKD enhancement proposal.

There's absolutely no reason that enhancements can't be incremental, addressing different use cases in separate enhancements.

> All that said, do you think moving all aspects of adding workers on day 2 into a separate proposal would be the best approach? I'm happy to make that change.

If we believe we can still make good decisions about the design with a smaller scope, I'd generally be in favour of the smaller scope. Now, maybe there's a chance we'll goof up the API design if we don't consider what the API for adding workers would be?

If you feel we can easily address agent-based adding of workers to Hive-installed clusters, feel free to keep it here. Maybe if the worker->cluster connection was more clear to me, I'd have a different perspective.

Reply:
As a first step, we removed adding workers from the Spoke cluster to itself.  This is not a priority for 4.8, and has some different details that I think are better suited for another enhancement (for example, no Hive, the involvement of a Machine actuator, etc.).  I do feel that the building blocks that we put forth will be a good fit there.

We have a couple of open issues around adding remote worker nodes from the Hub cluster.  The flow is described here but we're missing the part that is currently being discussed around worker vs. RWN roles.  I am inclined to move that to a separate enhancement as well.

> Oh wow. Maybe I was being dense, but it wasn't obvious to me we were talking about add workers specifically for clusters created by Hive.

The cluster doesn't need to be created by Hive, but we need some notion of a cluster that the new worker will belong to.  Hive can import/adopt an existing cluster, so the idea was to have a Hive ClusterDeployment regardless of how the cluster was installed, and then add workers in the same way.